---
title: "ggplot Exercise: scrnaseq"
output: html_document
---

## Load Libraries
You'll need the `tidyverse` and `ggExtra` packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggExtra)
```

## Required packages and datasets
This exercise uses a dataset package called `scRNAseq`, created by a former tidybiology TA. Install the package from his github repository. Load the library and the load the data using `data()`. *RENEMBER*: These data are MADE UP. Please therefore only use this dataset for tidybiology learning exercises!  
```{r}
#devtools::install_github("devangthakkar/scRNAseq", force = TRUE)
library(scRNAseq)
data(scRNAseq)
```

Take a quick look at the contents of this dataset before attempting the questions below
```{r}
glimpse(scRNAseq)
```

What is the distribution of transcript lengths for the genes in this dataset? Hint - the relevant variable is `transcript_length`
```{r}
#hint: this is a ggplot exercise
ggplot(scRNAseq) +
  geom_histogram(aes(x = transcript_length)) #for binned distributions; geom_density for smoothed
```

How many *unique* organs are represented in this dataset?
```{r}
scRNAseq %>% 
  distinct(organ)
```

Create a new data frame that only contains genes whose transcript length is less that 7500 bp long. Also discard all rows that contain "other" in the `organ` column. Call this new data frame `three_organ_df`
```{r}
three_organ_df <- scRNAseq %>% 
                    filter(transcript_length < 7500,
                           organ != 'other')
```

Create boxplots for transcript lengths for each organ type. Overlay the individual points on top of these boxplots
```{r}
ggplot(three_organ_df, aes(x = organ, y = transcript_length)) +
  geom_boxplot() +
  geom_point()
```

Give the previous plot appropriate labels for the x- and y-axes, and a title
```{r}
ggplot(three_organ_df, aes(x = organ, y = transcript_length)) +
  geom_boxplot() +
  geom_point() +
  labs(x = "Organ names", #labs gives you titles
       y = "Transcript Length",
       title = "Distribution of Transcript Length for Genes in each Organ")
```

Many of the points are obscuring each other. What's one way to "fix" this?
```{r}
ggplot(three_organ_df, aes(x = organ, y = transcript_length)) +
  geom_boxplot() +
  geom_jitter() + #jitters the points so they overlap less
  labs(x = "Organ names",
       y = "Transcript Length",
       title = "Distribution of Transcript Length for Genes in each Organ")
```

How would you make all the black points red?
```{r}
ggplot(three_organ_df, aes(x = organ, y = transcript_length)) +
  geom_boxplot() +
  geom_jitter(colour = 'red') + #easy color
  labs(x = "Organ names",
       y = "Transcript Length",
       title = "Distribution of Transcript Length for Genes in each Organ")
```

Plot the gene expression values for `cell_1` against `cell_2`. Make the points blue, change their size to 1.5, and adjust the transparency to 0.3. Then, add a diagonal reference line (45 degrees) that intersects with the intercept. Set the colour of this line to green.
```{r}
ggplot(three_organ_df, aes(x = cell_1, y = cell_2)) +
  geom_point(colour = 'blue',
             size = 1.5,
             alpha = 0.3) +
  geom_abline(intercept = 0,
              colour = 'green')
```

Change the plot above so that data points corresponding to different organ types have different colours
```{r}
ggplot(three_organ_df, aes(x = cell_1, y = cell_2)) +
  geom_point(aes(colour = organ),
             size = 1.5,
             alpha = 0.3) +
  geom_abline(intercept = 0,
              colour = 'green')
```

Split the plot above into three plots that correspond to the three organ types. Bonus: Have the label of each plot be in this format - organ: organ_name e.g. "organ: brain"
```{r}
ggplot(three_organ_df, aes(x = cell_1, y = cell_2)) +
  geom_point(aes(colour = organ),
             size = 1.5,
             alpha = 0.3) +
  facet_wrap(vars(organ), labeller = label_both, scales = "free_x") # ~organ also works in place of vars(organ)
```

Using the un-split version of the previous plot, add marginal histograms above and to the right of this plot. Hint - use the `ggMarginal()` function from the `ggExtra()` package
```{r}
p1 <- ggplot(three_organ_df, aes(x = cell_2, y = cell_1)) +
  geom_point(
    aes(colour = organ),
    size = 1.5,
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') 

ggExtra::ggMarginal(
  p = p1,
  type = 'histogram', #You can use other types of marginal plots here
  fill = 'gray'
)
```

Give the previous plot appropriate labels for the x- and y-axes, and a title. Also make the x- and y-axis text bold, and remove the colour legend
```{r}
p1 <- ggplot(three_organ_df, aes(x = cell_1, y = cell_2)) +
  geom_point(
    aes(colour = organ),
    size = 1.5,
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') +
  labs(x = "Cell 1 Transcriptome",
       y = "Cell 2 Transcriptome",
       title = "Comparing the Transcriptomes of Cell 1 and Cell 2") +
  theme(axis.text = element_text(face = "bold"),
        legend.position = "none")

ggExtra::ggMarginal(
  p = p1,
  type = 'histogram', #You can use other type of marginal plots here
  fill = 'gray'
)
```
