---
title: "ggplot Exercise: mitocarta"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("devtools")  #If you are loading the mitocarta packages for the first time you will likely need these
#library(devtools)
#devtools::install_github("hirscheylab/mitocarta")
library(tidyverse)
library(mitocarta)
```

```{r, include=FALSE}
#Load datasets we will use for the lecture

```

#lets glimpse at our data
```{r, include=FALSE}

```

#Basics of a ggplot code
Recall, the most basic form of the ggplot code has data, a geom, and an aesthetic that you map to data
```{r, eval=FALSE}
ggplot(data)+ #this will generate a blank plot for your data. At this point, ggplot knows what to plot, but not how.
  geom(mapping=aes(x, y)) #adding a geom tells ggplot how you would like to map your data
```

Let's make a simple scatterplot that compares `pgc_induction_score` to `coexpression_gnf_n50_score`. The `pgc_induction_score` variable represents the foldchange in transcript induction following overexpression of PGC1a known to induce mitochondrial biogenesis. The `coexpression_gnf_n50_score` variable represents the number of 50 nearest neighbors that are in Tmito (N50 score) for coexpression across mouse GNFv3 tissue atlas. Get the data dictionary at: https://rdrr.io/github/hirscheylab/mitocarta/src/R/data.R, or read the whole paper: https://mootha.med.harvard.edu/PubPDFs/Calvo2015NucleicAcids.pdf
```{r}
ggplot(data = mitocarta) + geom_point(mapping=aes(x = pgc_induction_score, y = coexpression_gnf_n50_score))
```

A note on ggplot style: while you can put the + at the beginning of the next line, it is generally put at the end of the previous line.

How does this chunk of code differ from the previous chunk?
```{r}
ggplot(data = mitocarta) + 
  geom_point(mapping=aes(x = pgc_induction_score, y = coexpression_gnf_n50_score))
```

## Geom refresher
When plotting your data, it is often helpful to take a glimpse at the data you intend to plot to know what kinds of variables you will be working with
```{r}

```

So now that you know your variable types, how do you know what geoms to use??
Use the following resources! 
  * https://rstudio.com/wp-content/uploads/2015/03/ggplot2-cheatsheet.pdf
  * https://www.data-to-viz.com/ 
  
## Changing Geoms
The following code generates a scatterplot of the mitocarta dataset. Change it to a boxplot
```{r}
#original scatter plot
ggplot(mitocarta) +
  geom_point(aes(x=hg19_chromosome,y=log(protein_length)))

#new boxplot

```
Sometimes you will run into errors indicating more information is needed for a plot or that you do not have the correct variable types. For more in depth information on the geoms, I find the ggplot2 reference page more helpful than the built in help pages
  * https://ggplot2.tidyverse.org/reference/index.html
  
Question- without running it, determine what the following code will do
```{r}
ggplot(mitocarta)+
  geom_jitter(aes(x=hg19_chromosome, y=log(protein_length)))+
  geom_boxplot(aes(x=hg19_chromosome, y=log(protein_length)))
```

## Aesthetics
Everything up to this point gets you a basic graph- but what about colors, shapes and overal style? 

There are 5 basic aesthetics you can can change
1. Color- changes the outline color of your datapoints
2. size
3. Shape
4. alpha- changes the transparency of each point
5. fill- changes the fill color of your points

## Change the aes
Add some different aes values to this graph.
```{r}
#original
ggplot(mitocarta) + 
	geom_point(aes(hg19_chromosome,log(protein_length)))

#new

```

## Practicing alpha
Adding alpha can help prevent over-plotting, which occurs when you cannot see all the points. Add alpha to the plot below to see if it fixes this problem.
```{r}
#original, no alpha
ggplot(mitocarta) + 
	geom_jitter(aes(hg19_chromosome,log(protein_length), color= rickettsia_homolog_score))

#updated, add alpha=0.5

```

#Global vs Local
In ggplot2, we have the options to set mappings globally or locally. Setting a mapping globally means to set those values in the original ggplot function.

Example:
  Earlier in class you made this graph:
```{r}
ggplot(mitocarta)+
  geom_jitter(aes(x=hg19_chromosome, y=log(protein_length)))+
  geom_boxplot(aes(x=hg19_chromosome, y=log(protein_length)))
```
  
However, if we map our x and y values in the ggplot function we find that we generate the same graph
```{r}
ggplot(mitocarta, aes(x=hg19_chromosome, y=log(protein_length)))+
  geom_jitter()+
  geom_boxplot()
```
This is because when you set the aes mappings in the orignal ggplot function you are setting the aes **globally**.
This means all the functions afterwards will inherit that mapping. So in our Example this means that both the jitter and boxplot geoms know to graph Cell Region by Protein Length

You can also set aes values **locally** within the geom function. Doing so will only change the values in that geom
```{r}
ggplot(mitocarta, aes(x=hg19_chromosome, y=log(protein_length)))+
  geom_jitter()+
  geom_boxplot(aes(color=hg19_chromosome))
```

Data can also be set locally or globally. For this example, let's filter our original data first using the dplyr::filter function. Starting with the original `mitocarta` dataframe, filter protein_length to only those that are > 6000. Finally, save this new object as `large_prot` for large proteins.
```{r}

```

Now, that we have a new object, we can practice combining global and local aesthetics. Let's label only the proteins identified in our large_prot data by setting data locally in a new geom
```{r}
#original plot
ggplot(mitocarta, aes(x=hg19_chromosome, y=log(protein_length)))+
  geom_jitter()+
  geom_boxplot(aes(color=hg19_chromosome))

#new plot with geom_label

```
You notice we have to indicate the new dataset, but because it has the same x and y values, we did not need to set those mappings.

## Labels and Legends
You've likely noticed that up until this point ggplot has labeled axes, but not necessarily in a very pleasing manner.
We can change these settings in our graph using the labs() function. 

Let's start by simply changing the x-axis label. 
```{r}
#paste the code from the last code chunk, and add a new labels layer
ggplot(mitocarta, aes(x=hg19_chromosome, y=log(protein_length)))+
  geom_jitter()+
  geom_boxplot(aes(color=hg19_chromosome))+
	geom_label(data=large_prot, aes(label= symbol))+
	labs(x= "Chromosome")
```

Now, seeing as we have a pretty explanatory legend, we could try to get rid of it. This becomes especially useful when ggplot gives you legends that don't make sense to show. There are sveral ways to remove legends. Try to find one using google. 
```{r}

```

## Finalize plot
Add appropriate x and y axis titles to the graph, as well as a title
```{r}

```

## Faceting
Faceting allows you to create multiple graphs side by side in one panel. Especially useful when you want to see the data together, but not on top of eachother  
  
For example:  
```{r}
ggplot(mitocarta)+
  geom_point(aes(x=hg19_chromosome, y=log(protein_length)))+
  facet_grid(cols = vars(mito_domain_score))
```
  
Run the following code to determine how they differ  
```{r}
q <- ggplot(mitocarta) + geom_point(aes(x = hg19_chromosome, y = log(protein_length)))
q + facet_grid(cols = vars(mito_domain_score))
q + facet_grid(rows = vars(mito_domain_score))
q + facet_wrap( ~ mito_domain_score)
```
  
You can even add more information to your faceting by adding another variable to facet on. Try facet wrapping q on `rickettsia_homolog_score` and `mito_domain_score`
```{r}

```

## Coordinate Flipping
Coordinate flipping is an especially useful tool when your x-axis includes characters that overlap. Coordinate flipping keeps your mappings and aes the same, but flips coordinates so that your x-axis displays on the Y-axis instead. This allows any names included in the x-axis titles to be read more easily
```{r}
#original
ggplot(mitocarta) + 
	geom_jitter(aes(hg19_chromosome,log(protein_length), color= rickettsia_homolog_score))

#add a coord_flip() layer

```

## Themes
You can change almost everything you see on your chart, but a lot of the things you may look to change are part of the "theme"

Here we are going to change some features about our title text
```{r}
ggplot(mitocarta) + 
	geom_jitter(aes(hg19_chromosome,log(protein_length), color= rickettsia_homolog_score))+
  coord_flip()+
  labs(x = "Chromosome", y = "Protein Length (log)", title = "Lengths of proteins encoded by each Chromosome")+
  theme(
     plot.title = element_text(face = "bold", size = 12)
    ) 
```

Adding specific elements to the `theme()` function allows you to change just about anything on your plot. Using the last plot, let's add another customization. Change the aesthetics of our legend box to set the fill to be "gray" and the color to be "black". Get resourceful (i.e. google) to determine how to do this. Hint: you want to change the legend.background
```{r}

```

Finally, let's change the legend postion. Again copy the plot from the last code chunk and pin the legend.position to the "bottom".
```{r}

```

Pre-Set themes also exist as an easy way to change the entire theme of your graph quickly. They can also be combined with custome theme settings
```{r}
#original
ggplot(mitocarta) + 
	geom_jitter(aes(hg19_chromosome,log(protein_length), color= rickettsia_homolog_score))+
  coord_flip()+
  labs(x = "Chromosome", y = "Protein Length (log)", title = "Lengths of proteins encoded by each Chromosome")+
  theme_bw()

#new theme

```

## Saving plots using ggsave
If you make a plot there are a few ways to save it, though the simplest is to use ggsave. 

For example, choose one of the plots you made earlier:
```{r}
#pick any plot from above and place here

```

You got all the geoms and aesthetics right and now you want to save it. All you have to do is run the following code
```{r}
ggsave("ggsaveexample.png")
```
You can change the type of file you save or even the size in inches. Add arguments to ggsave to save the image as a pdf (not png) and set the width and height to 6. 
```{r}

```

Where does it save? You can find out by typing `getwd()` which "GETs" you the "Working Directory. You can also pass a `path` argument to ggsave to save the plot in a specific spot.
```{r}

```

